### ISSUE TITLE
There should be a spacing option for formatting tuple syntax

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#10587]
There are options for spacing in parenthesis syntax in various contexts, but not tuples.

![image](https://cloud.githubusercontent.com/assets/12466233/14541294/1526cbf8-023e-11e6-99ba-3568e6730943.png)

**Comment by jcouv** [comment]
Assigning to Dustin as I don't think this involves compiler.

**Comment by DustinCampbell** [comment]
Removing assignee so we can look at this in triage.

**Comment by Pilchie** [comment]
Let's wait for a bit to see what emerges here.

**Comment by zeroskyx** [comment]
Any news on the topic? Having the tuple syntax as one of the most exciting new C# 7 features, not being able to get it formatted is a major drawback, especially for people from other languages we want to get excited.

**Comment by Pilchie** [comment]
@zeroskyx - Can you describe what options you'd like to see?  How do you *want* to format tuples?

**Comment by zeroskyx** [comment]
Basically to make tuples a consistent experience with lambda expressions and parameter declarations:
(int One, string Two) DoSomething()
( int One, string Two ) DoSomething()

Simply spacing after the opening and before the closing bracket ;)

**Comment by DustinCampbell** [comment]
Note that tuples are not the same as parameter declarations, they are types. We don't offer spacing options inside of generic type arguments either.

**Comment by Pilchie** [comment]
Also - @zeroskyx Are you saying you want to format them that way because you already format parameter lists that way, or just that you want consistency.  (Note that we don't have a specific option for lambda's either)

**Comment by zeroskyx** [comment]
I'm mostly interested in consistency that allow developers to customize the great formatting experience to their wishes.
Honestly I never noticed that generics can't be formatted -- maybe an option regarding spacing would be great there as well.
Regarding the tuples it just looks strange that spacing can be customized for every other aspect but not here.
It is also interesting that there is no option for lambdas since they format "correctly" in my case -- does the formatter inherit spacing from another position? Maybe re-using this for tuples would suffice.

**Comment by DustinCampbell** [comment]
In general, we try not to add formatting options unless there is a real desire/need for them. The fact that you'd never noticed that there weren't spacing for generic type parameter or argument lists is a good example of why they're not needed. In practice, we've found that adding formatting options just to be "completionist" can have a detrimental effect by presenting the user with a wall of options such that they can't find the option that they want, or they have to change dozens of options to get the desired behavior. Also, we've found that most users never touch formatting options or just choose one or two options (like brace formatting).

Inheriting the same behavior for tuples as parameter lists isn't appropriate because... well... they aren't parameter lists. We expect that many users will want their parameter and argument lists to format differently than tuples since they appear in different places in code. Also, there's the problem of tuple literals which look much different that tuple types.

**Comment by zeroskyx** [comment]
I totally agree with you for generics. However tuples/ return parameters pretty much behave/ look like lambda parameters:

var Lambda = () => 16;
var Lambda = a => 16;
var Lambda = ( a, b ) => 16;

void DoSomething();
int DoSomething();
( int, int ) DoSomething();

I don't know which parameter controls the spacing in the third lambda. Isn't re-using that parameter for tuples as well straight forward?
Is there any reason why one should be able to (implicitly) format lambdas while tuples cannot be changed and pretty much look alike?

So if creating a new setting for tuples is not what you want, please at least give us the possibility for a consistent style by aligning the formatting with lambdas.

**Comment by DustinCampbell** [comment]
To be clear, tuples can appear in a lot more places than parameter lists can. Also, we have separate formatting options for parameter lists and arguments lists, and I don't think tuples fit cleanly into either of those rules.

Consider the following examples of ways in which tuples can appear in code.

1. Local variable declaration as a type
    ```C#
    (int x, int y) p = GetPoint();
    ```

2. Deconstruction assignment (assigning to new variables)
    ```C#
    (int x, int y) = GetPoint();
    ```

3. Type-inferred local declaration
    ```C#
    var (x, y) = GetPoint();
    ```

4. Deconstruction assignment (assigning to other variables)
    ```C#
    int x, y;
    (x, y) = GetPoint();
    ```

5. As a tuple literal
    ```C#
    var p = (19, 23);
    ```

6. As a tuple literal (with named elements)
    ```C#
    var p = (x: 19, y: 23);
    ```

7. In a foreach statement as a type
    ```C#
    foreach ((int x, int y) p in points) { }
    ```

8. In a foreach statement (type-inferred)
    ```C#
    foreach (var (x, y) in points) { }
    ```

9. In a type-cast
    ```C#
    var p = ((long, long))GetPoint();
    ```

10. As a generic type argument
    ```C#
    var l = new List<(int x, int y)>();
    ```

11. As the return type of a member
    ```C#
    (int x, int y) GetPoint() => (19, 23);
    ```

12. A nested tuple as the return type of a member
    ```C#
    (int id, (string first, string last) name) GetUser() => (42, ("Dustin", "Campbell"));
    ```

13. As the type of parameter in a member
    ```
    void SetPoint((int x, int y) p) { }
    ```

My point is that tuples are pretty special and simply making them inherit the same formatting settings as parameter lists would probably not produce the results you hope for once you start using them throughout your code. If we decide to provide spacing rules for tuples, there should be (at a minimum) different rules for tuple types and tuple literals, and probably different rules for tuples that appear in declarations. However, as soon as we go down that road, there's a lot of other nuance to consider and design.

**Comment by danielkraffner** [comment]
I would love to have the spacing for tupels as well.The various locations where they appear do not require  various rules imho. For example, we have Spacing after Method names. Affects all methods, it's not like

- private method in class
- protected method in an enum
- Static method used as Extension Methods
-  ....

Here one clearly says: Why would you use different spacings for "that"? And I think the same would be true for Tupels.

**Comment by DustinCampbell** [comment]
@realdognose: Note that we *do* have different rules for methods. For methods, we allows different spacing rules for method declarations vs. methods calls.

**Comment by Pilchie** [comment]
@realdognose What options would you want?  What would you set them to?

**Comment by Thaina** [comment]
@jcouv By anything at all it should obey `SpaceBeforeComma` and `SpaceAfterComma` at the very least

**Comment by Thaina** [comment]
What? Why I have unassigned someone there?

**Comment by Neme12** [comment]
Maybe parameter list setting should affect tuple types and argument list settings should affect tuple expressions?

**Comment by CyrusNajmabadi** [comment]
> Maybe parameter list setting should affect tuple types and argument list settings should affect tuple expressions?

I would be in favor of this.

Note: as the IDE moves more to have analyzers that check and enforce formatting, we have to be very careful.   A team may have multiple versions of VS on their machine, and this could cause conflicts.  We're not there yet.  But presumably at some point we'll need to have to consider a breaking-change-like policy for our editorconfig/analyzers here as well.

**Comment by CyrusNajmabadi** [comment]
Would like to bring to next meeting.  There haven't been any actual requests for this (except to ask for consistency in options). However, without an actual real user need for this, i think we can just close this.

**Comment by iamjplant** [comment]
My team would like this!  That fact that it's a tuple under the hood feels somewhat irrelevant.  For our purposes, it looks like parenthesis, and we desire more spacing in there.  As it stands, the lack of parenthesis makes the code look odd to us.

**Comment by CyrusNajmabadi** [comment]
Closing out due to minimal community feedback here.
